src/content/Kentico.Kontent.Boilerplate/Caching/Webhooks/CacheManager.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Kentico.Kontent.Delivery;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Kentico.Kontent.Boilerplate.Caching.Webhooks
+{
+    public class CacheManager : ICacheManager, IDisposable
+    {
+        private readonly IMemoryCache _memoryCache;
+        private readonly CacheOptions _cacheOptions;
+        private readonly ConcurrentDictionary<string, SemaphoreSlim> _createLocks = new ConcurrentDictionary<string, SemaphoreSlim>();
+        private readonly ConcurrentDictionary<string, object> _dependencyLocks = new ConcurrentDictionary<string, object>();
+
+
+        public CacheManager(IMemoryCache memoryCache, IOptions<CacheOptions> cacheOptions)
+        {
+            _memoryCache = memoryCache;
+            _cacheOptions = cacheOptions.Value ?? new CacheOptions();
+        }
+
+        public async Task<T> GetOrAddAsync<T>(string key, Func<Task<T>> valueFactory, Func<T, IEnumerable<string>> dependenciesFactory, Func<T, bool> shouldCache = null)
+        {
+            if (TryGet(key, out T entry))
+            {
+                return entry;
+            }
+
+            var entryLock = _createLocks.GetOrAdd(key, _ => new SemaphoreSlim(1, 1));
+            try
+            {
+                await entryLock.WaitAsync();
+
+                if (TryGet(key, out entry))
+                {
+                    return entry;
+                }
+
+                var value = await valueFactory();
+
+                // Decide if the value should be cached based on the response
+                if (shouldCache != null && !shouldCache(value))
+                {
+                    return value;
+                }
+
+                // Set different timeout for stale content
+                var valueCacheOptions = new MemoryCacheEntryOptions();
+                if (value is AbstractResponse ar && ar.HasStaleContent)
+                {
+                    valueCacheOptions.SetAbsoluteExpiration(_cacheOptions.StaleContentTimeout);
+                }
+                else
+                {
+                    valueCacheOptions.SetSlidingExpiration(_cacheOptions.DefaultTimeout);
+                }
+
+                var dependencies = dependenciesFactory?.Invoke(value) ?? new List<string>();
+                var dependencyCacheOptions = new MemoryCacheEntryOptions { Priority = CacheItemPriority.NeverRemove };
+                foreach (var dependency in dependencies)
```

Comment:
Could the code be simplified by letting the `MemoryCache` instance manage dependencies? You already have a collection of dependency keys, so you can create dummy cache entries first (using the `Add` method) and then use the `CreateCacheEntryChangeMonitor` method to create a single change monitor for all dependencies.

Reply:
As discussed in person, the `MemoryCache` implementation in ASP.NET Core 3.0 has a new recommended way to work with dependencies. For details see https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0